File: src/spsc/msg.py
'''
Created on Aug 25, 2009

@author: Sergei Romanenko
'''

# MSG = Most Specific Generalization

from sll_language import *
from algebra import *

class Gen(object):

    def __init__(self, exp, substA, substB):
        self.exp = exp
        self.substA = substA
        self.substB = substB

    def __str__(self):
        assocA = list(self.substA.items())
        assocA.sort()
        assocA_s = ",".join(["%s=%s" % a for a in assocA])
        assocB = list(self.substB.items())
        assocB.sort()       
        assocB_s = ",".join(["%s=%s" % a for a in assocB])
        return "%s =>> {%s}{%s}" % (self.exp, assocA_s, assocB_s)

class MSGBuilder(object):

    def __init__(self, nameGen):
        self.nameGen = nameGen
        self.exp = None
        self.subst = None
        self.noProgress = None

    def build(self, e1, e2):
        vname = self.nameGen.freshName()
        self.exp = Var(vname)
        self.subst = dict([(vname, (e1, e2))])
        while True:
            self.noProgress = True
            self.mergeSubexp()
            if self.noProgress:
                self.commonFunctor()
            if self.noProgress:
                break
        subst1 = dict()
        subst2 = dict()
        for (vname, (e1, e2)) in self.subst.items():
            subst1[vname] = e1
            subst2[vname] = e2
        result = Gen(self.exp, subst1, subst2)
        self.exp = None
        self.subst = None
        return result

    def commonFunctor(self):
        for (vname, (e1, e2)) in self.subst.items():
            if e1.hasTheSameFunctorAs(e2):
                self.noProgress = False
                ns = self.nameGen.freshNameList(len(e1.args))
                vs = [Var(x) for x in ns]
                self.exp = self.exp.applySubst({vname : e1.cloneFunctor(vs)})
                del self.subst[vname]
                self.subst.update(zip(ns, zip(e1.args, e2.args)))
                return

    def aMergeableKeyPair(self):
        for i in self.subst.keys():
            for j in self.subst.keys():
                if i < j and self.subst[i] == self.subst[j]:
                    return (i, j)
        return None               

    def mergeSubexp(self):
        ij = self.aMergeableKeyPair()
        if not ij:
            return
        i = ij[0]
        j = ij[1]
        self.noProgress = False
        self.exp = self.exp.applySubst({i : Var(j)})
        del self.subst[i]

File: src/main/scala/spsc/MSG.scala
package spsc

import Algebra._
case class Gen(t: Term, m1: Map[Var, Term], m2: Map[Var, Term])
object MSG {
  def msg(t1: Term, t2: Term): Gen = {
    val v = freshVar()
    var g = Gen(v, Map(v -> t1), Map(v -> t2))
    var exp = g.t
    do {exp = g.t; g = commonSubst(commonFun(g))} while (exp != g.t)
    g
  }
  
  def commonFun(g: Gen): Gen = {
    for (v <- g.m1.keys) (g.m1(v), g.m2(v)) match {
      case (e1:CFG, e2:CFG) if shellEq(e1, e2) =>
        val vs = e1.args map freshVar
        val t = applySubst(Map(v -> e1.replaceArgs(vs)), g.t)
        return Gen(t, g.m1 - v ++ vs.zip(e1.args), g.m2 - v ++ vs.zip(e2.args))        
      case _ =>
    }
    g
  }
  
  def commonSubst(gen: Gen): Gen = {
    for ((v1, e1) <- gen.m1; (v2, e2) <- gen.m1)
      if ((v1 != v2 && e1 == e2) && (gen.m2(v1) == gen.m2(v2)))
        return Gen(applySubst(Map(v1 -> v2), gen.t), gen.m1 - v1, gen.m2 - v1)
    gen
  }
}
